File: src/utils/request.rs
use std::{net::TcpStream, io::{Read, Write}};
use serde::{Serialize, Deserialize};
use serde_json;

use super::{inputs, response::Response};

#[derive (Serialize, Deserialize)]
pub enum RequestType{
	CreateTask, UpdateTask, ListTasks, DeleteTask, DeleteAllTasks,
	CreateClient, UpdateClient, GetClient, DeleteClient,
}

#[derive (Serialize, Deserialize)]
pub struct Request {
	pub req: RequestType,
	pub cid: Option<String>,
	pub task: Option<String>,
	pub body: Option<String>
}

impl Request {
    pub fn from_string(data: String) -> Self {
        serde_json::from_str(data.as_str()).unwrap()
    }
    pub fn to_string(&self) -> String {
        serde_json::to_string(&self).unwrap()
    }
}

pub fn make_request(mut stream: &TcpStream, req: RequestType, mask: u8) -> Response {
	let inputs = inputs::get(mask);
	
	let req = Request {
		req,
		cid:  inputs.0,
		task: inputs.1,
		body: inputs.2
	};
	
	match stream.write(req.to_string().as_bytes()) {
		Ok(_) => {},
		Err(e) => {
			return Response::Error(e.to_string()) 
		},
	} 

	let mut res = String::new();
	match stream.read_to_string(&mut res) {
		Ok(_) => Response::from_string(res),
		Err(e) => Response::Error(e.to_string()),
	}
}

File: src/others/request.py
import json
from enum import IntEnum
from time import sleep

from .inputs import get as get_inputs
from .response import Response, ResponseType

DATA_PAYLOAD = 4096

class RequestType(IntEnum):
	CreateTask = 1
	UpdateTask = 2 
	ListTasks = 3
	DeleteTask = 4
	DeleteAllTasks = 5

	CreateClient = 6
	UpdateClient = 7
	GetClient = 8
	DeleteClient = 9

class Request():
	def __init__(self, req: RequestType, cid, task, body):
		self.req = req.value
		self.cid = cid
		self.task = task
		self.body = body

	def to_string(self):
		return json.dumps(self.__dict__)
	
	staticmethod
	def from_string(data):
		obj = json.loads(data)
		req = RequestType(int(obj['req']))
		return Request(req, obj['cid'], obj['task'], obj['body'])  


def make_request(stream, req, mask):
	cid, task, body = get_inputs(mask)
	request = Request(req, cid, task, body)
	
	print(f'Sending request...')
	try:
		str_request = bytes(request.to_string(), 'utf-8')
		stream.send(str_request)
		while True:
			sleep(1)
			response = stream.recv(DATA_PAYLOAD)
			if response:
				break
		return Response.from_string(response.decode('utf-8'))
	except Exception as e: 
		return Response(ResponseType.Error, str(e))